src/sentry/utils/event_frames.py

```diff
@@ -1,9 +1,50 @@
 from __future__ import annotations
 
-from typing import Any, Callable, Mapping, MutableMapping, Sequence
+from copy import deepcopy
+from typing import Any, Callable, Mapping, MutableMapping, Optional, Sequence, Tuple, cast
 
 from sentry.utils.safe import PathSearchable, get_path
 
+FrameMunger = Callable[[str, MutableMapping[str, Any]], bool]
+
+
+def java_frame_munger(key: str, frame: MutableMapping[str, Any]) -> bool:
+    if frame.get("filename") is None or frame.get("module") is None:
+        return False
+    if "/" not in str(frame.get("filename")) and frame.get("module"):
+        # Replace the last module segment with the filename, as the
+        # terminal element in a module path is the class
+        module = frame["module"].split(".")
+        module[-1] = frame["filename"]
+        frame[key] = "/".join(module)
+        return True
+    return False
+
+
+PLATFORM_FRAME_MUNGER: Mapping[str, FrameMunger] = {"java": java_frame_munger}
+
+
+def munged_filename_and_frames(
+    platform: str, data_frames: Sequence[Mapping[str, Any]], key: str = "munged_filename"
+) -> Optional[Tuple[str, Sequence[Mapping[str, Any]]]]:
+    """
+    Applies platform-specific frame munging for filename pathing.
+
+    Returns the key used to insert into the frames and a deepcopy of the input data_frames with munging applied,
+    otherwise returns None.
+    """
+    munger = PLATFORM_FRAME_MUNGER.get(platform)
+    if not munger:
+        return None
+
+    copy_frames: Sequence[MutableMapping[str, Any]] = cast(
+        Sequence[MutableMapping[str, Any]], deepcopy(data_frames)
+    )
+    frames_updated = False
+    for frame in copy_frames:
+        frames_updated |= munger(key, frame)
+    return (key, copy_frames) if frames_updated else None
```

Comment:
Decided to replace `supplement_filename` (which mutates the frame), with this which doesn't apply side-effects to the input. 

Instead, this returns a deep-copy of the frames if any munging happened.